### PR TITLE
Multicluster demo

### PR DESCRIPTION
**Pull request by allenlsy**
**Description**:

This is a continuation from the PR #3579 .

User can specify a list of Kubernetes contexts that the multicluster mode OSM will be installed, by updating the `MULTICLUSTER_CONTEXTS` field in the .env file. In the demo, bookstore is the service that can be accessed cross clusters. 

User can run the `./demo/run-osm-multicluster-demo.sh` command to install OSM on existing Kubernetes clusters. In the screenshot below, the OSMs are deployed on two local Kind clusters. The run this demo, first update `MULTICLUSTER_CONTEXTS` field in `.env` like this:

```
export MULTICLUSTER_CONTEXTS=(kind-osm-beta)
```

Then run the commands below:

```
KIND_CLUSTER_NAME=osm-alpha ./scripts/kind-with-registry.sh
KIND_CLUSTER_NAME=osm-beta ./scripts/kind-with-registry.sh
./demo/run-osm-multicluster-demo.sh
```

Screen Shot 2021-07-09 at 11 41 33 AM

**Affected area**:
| Functional Area            |     |
| -------------------------- | --- |
| New Functionality          | [X] |
| Documentation              | [ ] |
| Install                    | [ ] |
| CLI Tool                   | [ ] |
| Certificate Management     | [ ] |
| Control Plane              | [ ] |
| Ingress                    | [ ] |
| Egress                     | [ ] |
| Networking                 | [ ] |
| Observability              | [ ] |
| SMI Policy                 | [ ] |
| Sidecar Injection          | [ ] |
| Security                   | [ ] |
| Upgrade                    | [ ] |
| Tests                      | [ ] |
| CI System                  | [ ] |
| Demo                       | [X] |
| Performance                | [ ] |
| Other                      | [ ] |

Please answer the following questions with yes/no.

1. Does this change contain code from or inspired by another project?
    -   Did you notify the maintainers and provide attribution?

1. Is this a breaking change?

### CODE REVIEW

File: demo/run-osm-multicluster-demo.sh [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+
+# Build OSM binaries
+make build-osm
+
+for CLUSTER in "osm-alpha" "osm-beta"; do
+    KIND_CLUSTER_NAME=$CLUSTER
+    # shellcheck disable=SC2034
+    BOOKSTORE_CLUSTER_ID=$CLUSTER # this is used when deploying bookstore app
+    echo "Creating cluster named: $KIND_CLUSTER_NAME"
+    ./scripts/kind-with-registry.sh
```
Comment (snehachhabria) [inline demo/run-osm-multicluster-demo.sh:68]: Lets not force the creation of a cluster on kind. Can we assume the person running this script has two cluster already ? This can be a pre-req just like the run-osm-demo.sh script, where one can use a cluster on any env

File: demo/run-osm-multicluster-demo.sh [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+        --set=OpenServiceMesh.image.pullPolicy="$IMAGE_PULL_POLICY" \
+        --set=OpenServiceMesh.enableDebugServer="$ENABLE_DEBUG_SERVER" \
+        --set=OpenServiceMesh.enableEgress="$ENABLE_EGRESS" \
+        --set=OpenServiceMesh.deployGrafana="$DEPLOY_GRAFANA" \
+        --set=OpenServiceMesh.deployJaeger="$DEPLOY_JAEGER" \
+        --set=OpenServiceMesh.enableFluentbit="$ENABLE_FLUENTBIT" \
+        --set=OpenServiceMesh.deployPrometheus="$DEPLOY_PROMETHEUS" \
+        --set=OpenServiceMesh.envoyLogLevel="$ENVOY_LOG_LEVEL" \
+        --set=OpenServiceMesh.controllerLogLevel="trace" \
+        --set=OpenServiceMesh.featureFlags.enableMulticlusterMode="true" \        --timeout="$TIMEOUT" \
```
Comment (snehachhabria) [inline demo/run-osm-multicluster-demo.sh:155]:
```suggestion
        --set=OpenServiceMesh.featureFlags.enableMulticlusterMode="true" \
        --timeout="$TIMEOUT" \
```

File: Makefile [hunk 3]
```diff
@@ -170,6 +170,15 @@ kind-demo: export CTR_REGISTRY=localhost:5000
...
 
+.PHONY: kind-multicluster-demo
+kind-multicluster-demo:
+	./demo/run-osm-multicluster-demo.sh 
+
+.PHONY: kind-multicluster-reset
+kind-multicluster-reset:
+	kind delete cluster --name multicluster-osm-1
+	kind delete cluster --name multicluster-osm-2
+
```
Comment (snehachhabria) [inline Makefile:181]: This is not needed. We will add this when we plan to integrate with CI

File: pkg/providers/kube/client_test.go [hunk 4]
```diff
@@ -725,6 +724,7 @@ func TestListEndpointsForIdentity(t *testing.T) {
 			mockConfigurator := configurator.NewMockConfigurator(mockCtrl)
 			mockConfigController := config.NewMockController(mockCtrl)
 			mockConfigurator.EXPECT().GetFeatureFlags().Return(v1alpha1.FeatureFlags{EnableMulticlusterMode: false}).AnyTimes()
+			providerID := "provider"
```
Comment (snehachhabria) [inline pkg/providers/kube/client_test.go:727]: I am not sure why this change is a part of a demo setup

File: demo/run-osm-multicluster-demo.sh [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+    exit 1
+}
+
+# Check if Docker daemon is running
+docker info > /dev/null || { echo "Docker daemon is not running"; exit 1; }
+
+# Build OSM binaries
+make build-osm
+
+for CLUSTER in "osm-alpha" "osm-beta"; do
```
Comment (snehachhabria) [inline demo/run-osm-multicluster-demo.sh:63]: the name of the clusters can be specified in the .env, since this will differ from user to user

Comment (allenlsy) [reply demo/run-osm-multicluster-demo.sh]: Resolved. Configurable from .env

File: demo/run-osm-multicluster-demo.sh (hunk 1 above)
Comment (allenlsy) [reply demo/run-osm-multicluster-demo.sh]: Removed from this script. In the demo using Kind cluster, we can run this line manually.

File: demo/run-osm-multicluster-demo.sh [hunk 6]
```diff
@@ -0,0 +1,176 @@
+#!/bin/bash
+
+# Automated Multicluster Demo - Work in Progress
+# This script will deploy two kind clusters to the local machine.
+# Each cluster will have Open Service Mesh and the various Bookstore containers deployed to them.
+# Following Multicluster feature support - TBD on how we'll demonstrate multicluster access.
+#   Perhaps having the bookthief from one cluster access the bookstore in the other cluster.
```
Comment (snehachhabria) [inline demo/run-osm-multicluster-demo.sh:7]:
```suggestion
#   Perhaps having the bookbuyer from one cluster access the bookstore in the other cluster.
```

**Review by snehachhabria — APPROVED** [review]

File: pkg/utils/env.go [hunk 7]
```diff
@@ -11,3 +14,12 @@ func GetEnv(envVar string, defaultValue string) string {
...
 }
+
+// AppendClusterID is utilized to append extra DNS information to the FQDN if a clusterID is configured in the MeshConfig
+// <svc-name>.<svc-namespace> becomes <svc-name>.<svc-namespace>.svc.cluster.<clusterID>
+func AppendClusterID(baseFQDN, clusterID string) string {
+	if clusterID == "" {
+		return baseFQDN
+	}
+	return baseFQDN + fmt.Sprintf(".svc.cluster.%s", clusterID)
+}
```
Comment (shashankram) [inline pkg/utils/env.go:25]: This is only used in `books.go`, so move it there. Also seems out of place in `env.go`, if it needs to be reused since it's k8s specific, it should be in `pkg/k8s`.

**Review by shashankram — CHANGES_REQUESTED** [review]

File: pkg/utils/env.go (hunk 7 above)
Comment (allenlsy) [reply pkg/utils/env.go]: Moved to `pkg/k8s` as suggested.

File: .env.example [hunk 8]
```diff
@@ -24,6 +24,9 @@ export CTR_REGISTRY_PASSWORD=
 # Default: osm-system
 export K8S_NAMESPACE=osm-system
 
+# optional: name of Kubernetes contexts that OSM will be deployed in multicluster mode
+# export MULTICLUSTER_CONTEXTS=()
```
Comment (shashankram) [inline .env.example:28]:
```suggestion
# export MULTICLUSTER_CONTEXTS=
```

**Review by shashankram — APPROVED** [review]

File: .env.example (hunk 8 above)
Comment (allenlsy) [reply .env.example:28]: I want to indicate that this by default is an empty list. It will be used in for loop below

File: demo/run-osm-multicluster-demo.sh [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+    # Push to registry - needs to happen after registry creation
+    make docker-push
+
+    # Registry credentials
+    ./scripts/create-container-registry-creds.sh "$K8S_NAMESPACE"
+
+
+    # Deploys Xds and Prometheus
+    echo "Certificate Manager in use: $CERT_MANAGER"
+    if [ "$CERT_MANAGER" = "vault" ]; then
```
Comment (draychev) [inline demo/run-osm-multicluster-demo.sh:109]: We could remove this block. We would not need optionality for the purposes of the demo, since the demo is supposed to show whether/how multicluster OSM works. (One would assume if this works w/ one CA it will work with another.)
